[FancyZones] Configurable sensitivity radius

remyblok opened the pull request:
## Summary of the Pull Request
In the current version the FancyZones use a const of 20px when determining if two zones should be highlighted/window should be placed in two zones. These 20px are not very much on high resolution displays. This Pull request changes the const to a configurable setting per layout.

## PR Checklist
* [x] Applies to #6437 
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #6437 

## Info on Pull Request
- Add Setting the in FancyZone Editor for setting the Sensitivity radius
- Update XAML in editor to de-duplicate some code 
- Show settings (also the spacing setting) for both default and custom layouts
- Set default of 20px when setting is not yet present in json (i.e. upgrade scenario)

## Validation Steps Performed

Upgrade scenario:
- Have FancyZones configured
- Start new version
- JSON settings adds default 20 px for "editor-sensitivity-radius"

Usage:
- Open the FancyZone editor 
- New setting available "Space inside zone to highlight adjacent zone"
- Set the value to some higher value
- Apply setting
- Drag a window, using FancyZones, in between zones
- Both zones light up
- Now move away from the zone's edge. and see that the other zone now remains hightlighted for the ammount of pixels you set

enricogior commented:
Hi @remyblok 
thanks for the contribution.
Users with multiple monitors may run into the same issue this PR fixes, but in reverse order: having a good radius on a 4K monitor but too big on a low DPI monitor.
Adding a DPI multiplier of the radius sensitivity would make it consistent regardless on which monitor is applied.
Makes sense?

remyblok commented:
Hi @enricogior,
I'm not sure that adding the DPI there would be OK. The radius setting is per device, So basically each monitor gets its own radius anyway. 
I can image that determining the default is based on the DPI though. Then when a new monitor is plugged in it get the 20px * DPI multiplier to start with. 
But I would only apply to new users though. I thing it would be best to keep the 20 as default when upgrading the json settings, because users are now used to that already.
Do you agree?

enricogior commented:
@remyblok 
I want to make sure users will not be confused on how to use the option.
Having to set different values for each monitor is not ideal in my opinion, but I guess we can live with that since it's still an improvement over the current implementation, and it's similar to how the space around zones work (even if we have several negative feedback about the space around zones behavior, and we may change it in the Editor v2).

We should rename `Space inside zone to highlight adjacent zone` to something like `Distance to highlight adjacent zone` because this parameter is not really about the space inside the zone, but the distance between zones and because the option description should be as simple as possible.

remyblok commented:
I agree that users should not get confused. Would it help to add a visual indicator in the editor, like a dotted line where the distance is. Then when you change the value it is visible what has changed and how big the size in on your particular monitor. Is that a good idea?

I've updated the text and also moved it to the Resources.

Regarding documentation. The Wiki should get an update. I guess I need to create a separate pull request for the wiki parts, or...?

SeraphimaZ commented on src/modules/fancyzones/tests/UnitTests/ZoneSet.Spec.cpp:
```diff
@@ -26,7 +26,7 @@ namespace FancyZonesUnitTests
                 Assert::AreEqual(S_OK, hres);
 
-                ZoneSetConfig m_config = ZoneSetConfig(m_id, m_layoutType, Mocks::Monitor());
+                ZoneSetConfig m_config = ZoneSetConfig(m_id, m_layoutType, Mocks::Monitor(), 20);
```
It would be better to add a constant instead of magic numbers.

SeraphimaZ commented on src/modules/fancyzones/lib/JsonHelpers.cpp:
```diff
@@ -431,8 +433,8 @@ namespace JSONHelpers
-            result.data.zoneCount = static_cast<int>(
-                device.GetNamedNumber(NonLocalizable::EditorZoneCountStr));
+            result.data.zoneCount = static_cast<int>(device.GetNamedNumber(NonLocalizable::EditorZoneCountStr));
+            result.data.sensitivityRadius = static_cast<int>(device.GetNamedNumber(NonLocalizable::EditorSensitivityRadiusStr, 20));
```
Here would be better to add a constant. Let's create a namespace with constants in the `JsonHelpers.h` and add it there. Also, use this constant in the `FromJsonMissingSensitivityRadiusUsesDefault` test as the expected value.

SeraphimaZ reviewed (commented):
The code looks good and works as expected. A few minor changes should be added. Overall, nice work!

remyblok commented:
Thanks @SeraphimaZ for your positive feedback! I've updated the code to use a constant now.

enricogior commented:
@remyblok 
is it ready for final review? In case let's change it from "draft" to "open". Thanks.

enricogior reviewed (changes requested):
@remyblok 
I tested the code with clean settings and the default value for the sensitivity radius is set 0 instead of 20
![image](https://user-images.githubusercontent.com/3206696/93501072-3b22fe80-f915-11ea-94b3-b3d3e635c0e4.png)

enricogior commented on src/modules/fancyzones/lib/FancyZonesData.cpp:
```diff
@@ -445,9 +445,9 @@ bool FancyZonesData::SetAppLastZones(HWND window, const std::wstring& deviceId,
-                                                  .zoneIndexSet = zoneIndexSet };
+                                                    .zoneSetUuid = zoneSetId,
+                                                    .deviceId = deviceId,
+                                                    .zoneIndexSet = zoneIndexSet };
```
Can you restore the original formatting?

remyblok replied:
No clue why VS decided to reformat this part. But it is back to original!

enricogior commented on src/modules/fancyzones/lib/FancyZonesData.cpp:
```diff
@@ -555,112 +555,10 @@ void FancyZonesData::SaveFancyZonesData() const
+									appZoneHistoryFileName,
+									deviceInfoMap,
+									customZoneSetsMap,
+									appZoneHistoryMap);
```
Same here.

enricogior commented:
Tested, worked as expected. Thanks for the contribution!

crutkas commented:
@enricogior still lists a change was requested

enricogior commented:
@crutkas 
there was a formatting change still needed, now it's ready to be merged.

enricogior commented:
@remyblok 
if you wish to contribute with more fixes/features, it would be great. In case let's coordinate before starting some more work to make sure we are not already working on the same thing and if there are design aspects to clarify let's do it ahead of coding.
Thanks.
